***Heart Disease Prediction using Logistic Reggression model***

### Column Descriptions of attributes

age (Age of the patient in years)

sex (Male/Female)

cp chest pain type ([typical angina, atypical angina, non-anginal, asymptomatic])

trestbps resting blood pressure (resting blood pressure (in mm Hg on admission to the hospital))

chol (serum cholesterol in mg/dl)

fbs (if fasting blood sugar > 120 mg/dl)

restecg (resting electrocardiographic results)

thalach: maximum heart rate achieved

exang: exercise-induced angina (True/ False)

oldpeak: ST depression induced by exercise relative to rest

slope: the slope of the peak exercise ST segment

ca: number of major vessels (0-3) colored by fluoroscopy

thal: [normal; fixed defect; reversible defect]

target: the predicted attribute

***Importing dependencies***

In [43]:
import pandas as pd 
import numpy as np 
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import  accuracy_score

In [44]:
heart_dataset = pd.read_csv("heart.csv")
heart_dataset.head()


age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  slope  \
0   52    1   0       125   212    0        1      168      0      1.0      2   
1   53    1   0       140   203    1        0      155      1      3.1      0   
2   70    1   0       145   174    0        1      125      1      2.6      0   
3   61    1   0       148   203    0        1      161      0      0.0      2   
4   62    0   0       138   294    1        1      106      0      1.9      1   

   ca  thal  target  
0   2     3       0  
1   0     3       0  
2   0     3       0  
3   1     3       0  
4   3     2       0

The "target" field refers to the presence of heart disease in the patient. It is integer valued 0 = no disease and 1 = disease.

### Checking the information about data

In [45]:
heart_dataset.info

<bound method DataFrame.info of       age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  \
0      52    1   0       125   212    0        1      168      0      1.0   
1      53    1   0       140   203    1        0      155      1      3.1   
2      70    1   0       145   174    0        1      125      1      2.6   
3      61    1   0       148   203    0        1      161      0      0.0   
4      62    0   0       138   294    1        1      106      0      1.9   
...   ...  ...  ..       ...   ...  ...      ...      ...    ...      ...   
1020   59    1   1       140   221    0        1      164      1      0.0   
1021   60    1   0       125   258    0        0      141      1      2.8   
1022   47    1   0       110   275    0        0      118      1      1.0   
1023   50    0   0       110   254    0        0      159      0      0.0   
1024   54    1   0       120   188    0        1      113      0      1.4   

      slope  ca  thal  target  
0         2

### Looking for null or missing values in data

In [46]:
heart_dataset.isnull().sum()

age         0
sex         0
cp          0
trestbps    0
chol        0
fbs         0
restecg     0
thalach     0
exang       0
oldpeak     0
slope       0
ca          0
thal        0
target      0
dtype: int64

### Checking the data types of the attributes of data

In [47]:
heart_dataset.dtypes

age           int64
sex           int64
cp            int64
trestbps      int64
chol          int64
fbs           int64
restecg       int64
thalach       int64
exang         int64
oldpeak     float64
slope         int64
ca            int64
thal          int64
target        int64
dtype: object

### Knowing about statistical measures of data

In [48]:
heart_dataset.describe()

age          sex           cp     trestbps        chol  \
count  1025.000000  1025.000000  1025.000000  1025.000000  1025.00000   
mean     54.434146     0.695610     0.942439   131.611707   246.00000   
std       9.072290     0.460373     1.029641    17.516718    51.59251   
min      29.000000     0.000000     0.000000    94.000000   126.00000   
25%      48.000000     0.000000     0.000000   120.000000   211.00000   
50%      56.000000     1.000000     1.000000   130.000000   240.00000   
75%      61.000000     1.000000     2.000000   140.000000   275.00000   
max      77.000000     1.000000     3.000000   200.000000   564.00000   

               fbs      restecg      thalach        exang      oldpeak  \
count  1025.000000  1025.000000  1025.000000  1025.000000  1025.000000   
mean      0.149268     0.529756   149.114146     0.336585     1.071512   
std       0.356527     0.527878    23.005724     0.472772     1.175053   
min       0.000000     0.000000    71.000000     0.000000     0.000000   
25%       0.000000     0.000000   132.000000     0.000000     0.000000   
50%       0.000000     1.000000   152.000000     0.000000     0.800000   
75%       0.000000     1.000000   166.000000     1.000000     1.800000   
max       1.000000     2.000000   202.000000     1.000000     6.200000   

             slope           ca         thal       target  
count  1025.000000  1025.000000  1025.000000  1025.000000  
mean      1.385366     0.754146     2.323902     0.513171  
std       0.617755     1.030798     0.620660     0.500070  
min       0.000000     0.000000     0.000000     0.000000  
25%       1.000000     0.000000     2.000000     0.000000  
50%       1.000000     0.000000     2.000000     1.000000  
75%       2.000000     1.000000     3.000000     1.000000  
max       2.000000     4.000000     3.000000     1.000000

### Checking the distribution of target or label attribute

In [49]:
heart_dataset['target'].value_counts()

1    526
0    499
Name: target, dtype: int64

so as it is already mentioned that 1 refer to patient that has heart disease and 0 refer to a patient that has no heart 
disease so there are 526 are heart disease patients and 499 are not.

### Separating the features and target

In [50]:
x = heart_dataset.drop(columns=['target'], axis=1)

y = heart_dataset['target']

#### Features

In [51]:
x[:5]

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  slope  \
0   52    1   0       125   212    0        1      168      0      1.0      2   
1   53    1   0       140   203    1        0      155      1      3.1      0   
2   70    1   0       145   174    0        1      125      1      2.6      0   
3   61    1   0       148   203    0        1      161      0      0.0      2   
4   62    0   0       138   294    1        1      106      0      1.9      1   

   ca  thal  
0   2     3  
1   0     3  
2   0     3  
3   1     3  
4   3     2

#### Target

In [52]:
y[:10]

0    0
1    0
2    0
3    0
4    0
5    1
6    0
7    0
8    0
9    0
Name: target, dtype: int64

### Splitting the data into training and testing

In [53]:
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.2, stratify=y , random_state=2)

#### Looking for trainging and test data

In [54]:
#oringnal shape of x and y
print(x.shape, y.shape)

(1025, 13) (1025,)


In [55]:
#shape after splitting
print(x_train.shape, x_test.shape , y_train.shape, y_test.shape)

(820, 13) (205, 13) (820,) (205,)


### Training the model

In [56]:
model = LogisticRegression()

#### Fittinng the model over training data

In [57]:
model.fit(x_train,y_train)

c:\Users\Madina Computers\anaconda3\lib\site-packages\sklearn\linear_model\_logistic.py:814: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


LogisticRegression()

## Model Evaluation(predict and check the accuracy score)

#### Model Evaluation for training data

In [58]:
#prediction
training_data_prediction = model.predict(x_train)

#checking accuracy score
training_data_accuracy = accuracy_score(training_data_prediction, y_train)

print("This is the accuracy score for training data : ",training_data_accuracy)


This is the accuracy score for training data :  0.848780487804878


#### Model Evaluation for test data

In [59]:
#prediction
test_data_prediction = model.predict(x_test)

#checking accuracy score
test_data_accuracy = accuracy_score(test_data_prediction, y_test)

print("This is the accuracy score for test data : ",test_data_accuracy)


This is the accuracy score for test data :  0.8048780487804879


### Build a System for both a heart patient and not a heart patient

In [60]:
#get the values from user
input_values = (34,0,1,118,210,0,1,192,0,0.7,2,0,2)

#coverting the values into array(numpy array)

input_array = np.asarray(input_values)

#re-shapeing becouse we only want to predict for one instance as 0 or 1

re_shape_data = input_array.reshape(1,-1)

prediction = model.predict(re_shape_data)
print(prediction)

if prediction == 1:
    print("The patient has heart disease")
else:
    print("The patient has not heart disease")







[1]
The patient has heart disease


c:\Users\Madina Computers\anaconda3\lib\site-packages\sklearn\base.py:450: UserWarning: X does not have valid feature names, but LogisticRegression was fitted with feature names
  warnings.warn(


In [64]:
#get the values from user
input_values = (66,0,0,178,228,1,1,165,1,1,1,2,3)

#coverting the values into array(numpy array)

input_array = np.asarray(input_values)

#re-shapeing becouse we only want to predict for one instance as 0 or 1

re_shape_data = input_array.reshape(1,-1)

prediction = model.predict(re_shape_data)
print(prediction)

if prediction == 1:
    print("The patient has heart disease")
else:
    print("The patient has not heart disease")







[0]
The patient has not heart disease


c:\Users\Madina Computers\anaconda3\lib\site-packages\sklearn\base.py:450: UserWarning: X does not have valid feature names, but LogisticRegression was fitted with feature names
  warnings.warn(


***Note!*** 

#### You can also compare and confirm the output of our model with actual data for verification or justification.